File: 3-longest-substring-without-repeating-characters/3-longest-substring-without-repeating-characters.cs
public class Solution {
    public int LengthOfLongestSubstring(string s) {
        Dictionary<char,int> d = new Dictionary<char,int>();
        int max = 0;
        int i = 0;
        int start = 0;
        foreach (var c in s)
        {
            if(!d.ContainsKey(c))
            {
                d.Add(c,i);
            }
            else
            {
                if(d[c]<start)
                {
                    d[c]=i;
                }
                else
                {
                    max = max>(i-start)?max:(i-start);
                    start=d[c]+1;
                    d[c]=i;
                }
            }
            ++i;
        }
        max = max>(i-start)?max:(i-start);
        return max;
    }
}

File: 3-longest-substring-without-repeating-characters/3-longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        int start = 0;
        unordered_map<char,int> m;
        int end = 0;
        int result = 0;
        for(char c : s)
        {
            if(m.find(c)!=m.end() && m[c] >= start)
                start = m[c]+1;
            m[c]=end;
            ++end;
            result = max(result,end-start);
        }
        return result;
    }
};


/*
abccabcbb
start = 0

end = 2
p x
w x
w 2
k
e
w

*/